[FLINK-4469] [table] Add support for user defined table function in Table API & SQL

wuchong opened the pull request:
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.
- [x] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)
- [ ] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added
- [x] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

This PR introduces user-defined table functions for the Table and SQL API.
I will add documentation after this proposal is accepted. This is the general syntax so far:

In Java:

``` java
public class Split extends TableFunction<String> {
    public int eval(String str) {
        for (String s : str.split(" ")) {
            collect(s);
        }
    }
}

tableEnv.registerFunction("split", new Split());
// cross apply
Table result = table.crossApply("split(c) as (s)").select("c, s")
Table result = tableEnv.sql("SELECT MyTable.c, t.s FROM MyTable, LATERAL TABLE(split(c)) as t(s)")

// outer apply
Table result = table.outerApply("split(c) as (s)").select("c, s")
Table result = tableEnv.sql("SELECT MyTable.c, t.s FROM MyTable LEFT JOIN LATERAL TABLE(split(c)) as t(s)")
```

In Scala

``` scala
object Split extends TableFunction<String> {
  def eval(str: String): Unit = {
    str.split(" ").foreach(collect)
  }
}

// cross apply
val result = table.crossApply(Split('c) as ('s)).select('c, 's)
tableEnv.registerFunction("split", Split);
val result = tableEnv.sql("SELECT MyTable.c, t.s FROM MyTable, LATERAL TABLE(split(c)) as t(s)")

// outer apply
val result = table.outerApply(Split('c) as ('s)).select('c, 's)
tableEnv.registerFunction("split", Split);
val result = tableEnv.sql("SELECT MyTable.c, t.s FROM MyTable LEFT JOIN LATERAL TABLE(split(c)) as t(s)")
```

wuchong commented:
Hi @twalthr @fhueske  , it will be great if you can review this.

wuchong commented:
Because SQL doesn't have standard user-defined table function, we introduce MS SQL Server's `CROSS/OUTER APPLY` to support UDTF , see https://technet.microsoft.com/en-us/library/ms175156(v=sql.105).aspx .

NOTE: In order to implement the `CROSS/OUTER APPLY`, we need to modify Calcite's SQL Parser. So we include `Parser.jj` from Calcite source code and add some modifications (search FLINK in this file). 

See [this](https://docs.google.com/document/d/15iVc1781dxYWm3loVQlESYvMAxEzbbuVFPZWBYuY1Ek/edit#) for more UDTF design,

wuchong commented:
Hi @fhueske @twalthr ,  do you have any thoughts on this pull request ?

fhueske commented:
HI @wuchong, thanks for this feature. This looks really interesting :-)

I haven't looked at the PR in detail yet. However, I would like to avoid customizing the parser if possible. I haven't tested it but would it be possible to express `OUTER APPLY` and `CROSS APPLY` with Calcite's `LATERAL` tables?

`CROSS APPLY` corresponds to: 

```
SELECT MyTable.*, t.s  FROM MyTable, LATERAL TABLE(split(MyTable.a)) AS t(s)
```

`OUTER APPLY` could be:

```
SELECT MyTable.*, t.s  
  FROM MyTable LEFT JOIN LATERAL TABLE(split(MyTable.a)) AS t(s)
```

Adding `crossApply` and `outerApply` for the Table API is fine with me.

shaoxuan-wang commented:
Jack,
Giving this for the second thought, I feel we can just use LATERAL and
INNER/LEFT JOIN (which is already supported by Calcite) instead of
introducing "CROSS/OUTER APPLY" . Let us sync offline and see how to revise
the current PR.


On Mon, Oct 24, 2016 at 11:07 AM, Jark <notifications@github.com> wrote:

> Hi @fhueske <https://github.com/fhueske> @twalthr
> <https://github.com/twalthr> , do you have any thoughts on this pull
> request ?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/flink/pull/2653#issuecomment-255638756>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AI9yKgzYvailMK0PZM8ErLd70GlAfvChks5q3CEHgaJpZM4KZVsH>
> .
>

wuchong commented:
@fhueske I have updated this PR for the following changes.

1. Remove CROSS/OUTER APPLY support in SQL
2. Change Java Table API from `.crossApply("split(c)", "s")` to `.corssApply("split(c) as (s)")`
3. Make `ExpressionParser` can parse a string to `LogicalNode`

I think the PR is ready to be reviewed :)

fhueske commented:
Thanks for the update @wuchong! Will have a look at this PR.

sunjincheng121 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/TableEnvironment.scala:
```diff
@@ -152,21 +153,40 @@ abstract class TableEnvironment(val config: TableConfig) {
-        // register in SQL API
-        functionCatalog.registerSqlFunction(sf.getSqlFunction(name, typeFactory))
+    // register in SQL API
+    functionCatalog.registerSqlFunction(function.getSqlFunction(name, typeFactory))
```
"At here, function.getSqlFunction (name, typeFactory)" the "name" is user input,but at ScalarFunctionCall I see "scalarFunction.getSqlFunction (scalarFunction.toString, typeFactory)", which may not be we expect Right?

wuchong commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/functions/EvaluableFunction.scala:
```diff
@@ -0,0 +1,62 @@
+/*
```
I'm ok to move these methods to utils. Because they are only used in compile phase, and it will simplify TableFunction and ScalarFunction interface. What do you think @fhueske ?

fhueske replied:
+1 I think it's a good idea to move these methods.

wuchong commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/functions/TableFunction.scala:
```diff
@@ -0,0 +1,119 @@
+      name: String,
+      typeFactory: FlinkTypeFactory)
+    : SqlFunction = {
+    null
```
We need to create different `SqlFunction`s for every eval methods. So the `createSqlFunction(name: String, typeFactory: FlinkTypeFactory)` doesn't meet the requirement. I would like the move `createSqlFunction`  and `createSqlFunctions` to utils.

wuchong commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/TableEnvironment.scala:
```diff
@@ -152,21 +153,40 @@ abstract class TableEnvironment(val config: TableConfig) {
+      implicitly[TypeInformation[T]]
     }
+
+    val (fieldNames, fieldIndexes) = UserDefinedFunctionUtils.getFieldInfo(typeInfo)
```
It is possible. But I would like the move `createSqlFunction`  and `createSqlFunctions` to utils.

sunjincheng121 replied:
that’s good

wuchong commented:
Hi @sunjincheng121 ,  thank you for the reviewing. I will update the PR according to your comments.

fhueske commented:
Hi @wuchong, I did a first high-level pass over the PR. From what I've seen it looks really good and I do not expect that major changes are necessary 👍. Will do a more thorough review in the next days.

Thanks, Fabian

wuchong commented:
Sounds great. Thanks Fabian .

twalthr commented:
@wuchong thanks for the PR. I will also review it the next days.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/rules/dataSet/DataSetCorrelateRule.scala:
```diff
@@ -0,0 +1,89 @@
+      val convInput: RelNode = RelOptRule.convert(join.getInput(0), DataSetConvention.INSTANCE)
+      val right: RelNode = join.getInput(1)
+
+      def convertToCorrelate(relNode: RelNode, condition: RexNode): DataSetCorrelate = {
```
define `condition` as `Option[RexNode]` so we do not have to use `null`

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/rules/datastream/DataStreamCorrelateRule.scala:
```diff
@@ -0,0 +1,89 @@
+    val convInput: RelNode = RelOptRule.convert(join.getInput(0), DataStreamConvention.INSTANCE)
+    val right: RelNode = join.getInput(1)
+
+    def convertToCorrelate(relNode: RelNode, condition: RexNode): DataStreamCorrelate = {
```
define `condition` as `Option[RexNode]` so we do not have to use `null`

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/datastream/DataStreamCorrelate.scala:
```diff
@@ -0,0 +1,138 @@
+    traitSet: RelTraitSet,
+    inputNode: RelNode,
+    scan: LogicalTableFunctionScan,
+    condition: RexNode,
```
use `Option[RexNode]` for `condition`

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetCorrelate.scala:
```diff
@@ -0,0 +1,136 @@
+    traitSet: RelTraitSet,
+    inputNode: RelNode,
+    scan: LogicalTableFunctionScan,
+    condition: RexNode,
```
use `Option[RexNode]` for `condition`

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetCorrelate.scala:
```diff
@@ -0,0 +1,136 @@
+
+
+  override def computeSelfCost(planner: RelOptPlanner, metadata: RelMetadataQuery): RelOptCost = {
+    val rowCnt = metadata.getRowCount(getInput) + 10
```
Not that is would actually matter at the moment, but why are you adding a constant here. Shouldn't it be something like `* 1.5` instead?

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetCorrelate.scala:
```diff
@@ -0,0 +1,136 @@
+
+
+  override def translateToPlan(tableEnv: BatchTableEnvironment,
+                               expectedType: Option[TypeInformation[Any]]): DataSet[Any] = {
```
please indent similar as the constructor parameters of this class

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/datastream/DataStreamCorrelate.scala:
```diff
@@ -0,0 +1,138 @@
+
+
+  override def translateToPlan(tableEnv: StreamTableEnvironment,
+                               expectedType: Option[TypeInformation[Any]]): DataStream[Any] = {
```
please indent similar as the constructor parameters of this class.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/datastream/DataStreamCorrelate.scala:
```diff
@@ -0,0 +1,138 @@
+      config.getEfficientTypeUsage)
+
+    val inputDS = inputNode.asInstanceOf[DataStreamRel]
+      .translateToPlan(tableEnv, Some(inputRowType(inputNode)))
```
I think we can replace `Some(inputRowType(inputNode))` by `Some(TypeConverter.DEFAULT_ROW_TYPE)` (similar as in `DataSetAggregate.translateToPlan()`

Then we can also remove the method `FlinkCorrelate.inputRowType`.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetCorrelate.scala:
```diff
@@ -0,0 +1,136 @@
+      config.getEfficientTypeUsage)
+
+    val inputDS = inputNode.asInstanceOf[DataSetRel]
+      .translateToPlan(tableEnv, Some(inputRowType(inputNode)))
```
I think we can replace `Some(inputRowType(inputNode))` by `Some(TypeConverter.DEFAULT_ROW_TYPE)` (similar as in `DataSetAggregate.translateToPlan()`

Then we can also remove the method `FlinkCorrelate.inputRowType`.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/schema/FlinkTableFunctionImpl.scala:
```diff
@@ -0,0 +1,77 @@
+import org.apache.flink.api.common.typeutils.CompositeType
+import org.apache.flink.api.table.{FlinkTypeFactory, TableException}
+
+class FlinkTableFunctionImpl[T](val typeInfo: TypeInformation[T],
```
Please indent parameters as in `DataSetCorrelate`.

fhueske replied:
Please add a brief description of the class.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/validate/FunctionCatalog.scala:
```diff
@@ -47,13 +52,50 @@ class FunctionCatalog {
   }
 
+  /** Register multiple sql functions at one time. The functions has the same name. **/
+  def registerSqlFunctions(functions: Seq[SqlFunction]): Unit = {
```
Please check that all functions have the same name.

wuchong replied:
sure

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/validate/FunctionCatalog.scala:
```diff
@@ -47,13 +52,50 @@ class FunctionCatalog {
+          case Failure(e) => throw ValidationException(e.getMessage)
+        }
+      case _ =>
+        throw ValidationException("Unsupported table function.")
```
I think this exception message could be improved. It is throw if the registered method does not implement the `TableFunction` interface.

wuchong replied:
You are right.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/functions/EvaluableFunction.scala:
```diff
@@ -0,0 +1,62 @@
+      }
+
+    if (methods.isEmpty) {
+      throw new ValidationException(s"Table function class '$this' does not implement at least " +
```
If this method is also used for `ScalarFunction`, the exception message should be adapted accordingly.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/functions/TableFunction.scala:
```diff
@@ -0,0 +1,119 @@
+      name: String,
+      typeFactory: FlinkTypeFactory)
+    : SqlFunction = {
+    throw new UnsupportedOperationException("this method should not be called")
```
Why is this method not necessary for `TableFunction`?

wuchong replied:
Because we have to register `TableSqlFunction`s for every `eval` method. But the interface `createSqlFunction(name: String, typeFactory: FlinkTypeFactory)  : SqlFunction` only return one SqlFunction. 

I'm thinking about moving `createSqlFunction` to utils.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/rules/dataSet/DataSetCorrelateRule.scala:
```diff
@@ -0,0 +1,89 @@
+import org.apache.flink.api.table.plan.nodes.dataset.{DataSetConvention, DataSetCorrelate}
+
+/**
+  * parser cross/outer apply
```
Replace by "Rule to convert a LogicalCorrelate into a DataSetCorrelate.".

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/rules/datastream/DataStreamCorrelateRule.scala:
```diff
@@ -0,0 +1,89 @@
+import org.apache.flink.api.table.plan.nodes.datastream.{DataStreamCorrelate, DataStreamConvention}
+
+/**
+  * parser cross/outer apply
```
Replace by "Rule to convert a LogicalCorrelate into a DataStreamCorrelate.".

wuchong commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/functions/TableFunction.scala:
```diff
@@ -0,0 +1,119 @@
+
+/**
+  * Base class for a user-defined table function (UDTF). A user-defined table functions works on
+  * one row as input and returns multiple rows as output.
```
You are right.

wuchong commented on flink-libraries/flink-table/src/test/java/org/apache/flink/api/java/batch/UserDefinedTableFunctionITCase.java:
```diff
@@ -0,0 +1,201 @@
+
+
+	@Test
+	public void testUDTF() throws Exception {
```
sure

fhueske commented:
Maybe one more thing to add. @twalthr recently added some tooling to write unit tests that check the translation from SQL / Table API to `DataSetRel` and `DataStreamRel` nodes, i.e., exclude the final translation into DataSet and DataStream programs. See PR #2595.
A couple of ITCases could be replaced by such tests. We need of course a few end-to-end tests nonetheless, but maybe not as many as currently.

I also noticed that there are not tests that check for correct failures, e.g., check correct errors if a SQL query refers to a function in `FROM` that has not been registered or which is a ScalarFunction.

wuchong commented:
Hi @fhueske , you mentioned two ways to reduce IT cases. One is comparing the logical plans of two tables, this can reduce Java IT cases. Another is using `TableTestBase` tool to write unit tests, this can reduce both Java and SQL IT cases I think. 

So which one do you suggest ?  The latter one or both ?

fhueske commented:
I think we should convert the tests for the Java String-based Table API to check the logical plans of queries against the logical plan of a corresponding Scala Expression-based Table API query.

That leaves the Scala Table API and SQL tests for batch and streaming. I think some of those can be checked with @twalthr's test framework.

twalthr commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/functions/utils/UserDefinedFunctionUtils.scala:
```diff
@@ -162,24 +191,107 @@ object UserDefinedFunctionUtils {
+      } catch {
+        case ite: InvalidTypesException =>
+          throw new ValidationException(
+            s"Return type of table function '$this' cannot be " +
```
If this is used for both methods, than `table function` is wrong.

twalthr commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/functions/utils/UserDefinedFunctionUtils.scala:
```diff
@@ -162,24 +191,107 @@ object UserDefinedFunctionUtils {
+  def getFieldInfo(inputType: TypeInformation[_])
+  : (Array[String], Array[Int]) = {
+    val fieldNames: Array[String] = inputType match {
+      case t: TupleTypeInfo[_] => t.getFieldNames
```
Why don't you use `CompositeType` here?

wuchong replied:
Yes, you are right. We can use `CompositeType` here.

twalthr commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/functions/utils/UserDefinedFunctionUtils.scala:
```diff
@@ -162,24 +191,107 @@ object UserDefinedFunctionUtils {
+    * @tparam A The type of the TypeInformation.
+    * @return A tuple of two arrays holding the field names and corresponding field types.
+    */
+  def getFieldAttribute[A](inputType: TypeInformation[A])
```
Maybe merge this method and the method above?

twalthr commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/call.scala:
```diff
@@ -0,0 +1,169 @@
+  }
+}
+
+
```
Empty lines.

twalthr commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/call.scala:
```diff
@@ -0,0 +1,169 @@
+import scala.collection.JavaConversions._
+
+/**
+  * General expression for unresolved user-defined table function calls.
```
This is not an expression.

twalthr commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/FlinkCorrelate.scala:
```diff
@@ -0,0 +1,161 @@
+    val call = generator.generateExpression(rexCall)
+    var body = call.code +
+               s"""
+                  |scala.collection.Iterator iter = ${call.resultTerm}.getRowsIterator();
```
Until now we have pure Java generated code. Do we have to use Scala here?

wuchong replied:
You are right,  replace it by Java Iterator can work too.

twalthr commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/FlinkCorrelate.scala:
```diff
@@ -0,0 +1,161 @@
+      returnType, rowType.getFieldNames)
+
+    val input2NullExprs = input2AccessExprs.map(
+      x => GeneratedExpression("null", "true", "", x.resultType))
```
We use primitives where possible, so null might result in a compiler error for some types.

twalthr commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/FlinkCorrelate.scala:
```diff
@@ -0,0 +1,161 @@
+      returnType, rowType.getFieldNames)
+
+    val call = generator.generateExpression(rexCall)
+    var body = call.code +
```
I would put this into the multiline string.

twalthr commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/FlinkCorrelate.scala:
```diff
@@ -0,0 +1,161 @@
+  }
+
+  private[flink] def correlateOpName(rexCall: RexCall,
+                                   sqlFunction: TableSqlFunction,
```
In general we use only two spaces for indention.

twalthr commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/schema/FlinkTableFunctionImpl.scala:
```diff
@@ -0,0 +1,77 @@
+      case cType: CompositeType[T] =>
+        if (fieldNames.length != cType.getArity) {
+          throw new TableException(
+            s"Arity of type (" + cType.getFieldNames.deep + ") " +
```
deep=length?

wuchong replied:
No. `deep` is used to pretty print field names array such as `Array("a", "b", "c").deep.toString` will print "Array("a", "b", "c")" . 

I copied this from `FlinkTable`.

twalthr commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/table.scala:
```diff
@@ -611,6 +612,130 @@ class Table(
   }
 
   /**
+    * The Cross Apply returns rows form the outer table (table on the left of the Apply operator)
```
from

twalthr commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/table.scala:
```diff
@@ -611,6 +612,130 @@ class Table(
+  }
+
+  /**
+    * The Cross Apply returns rows form the outer table (table on the left of the Apply operator)
```
form=from

twalthr commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/validate/FunctionCatalog.scala:
```diff
@@ -47,13 +52,50 @@ class FunctionCatalog {
   /**
+    * Lookup table function and create an TableFunctionCall if we find a match.
+    */
+  def lookupTableFunction[T](name: String, children: Seq[Expression]): TableFunctionCall[T] = {
```
If `TableFunctionCall` would be an expression, we could merge this function with the function below.

twalthr commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/validate/FunctionCatalog.scala:
```diff
@@ -47,13 +52,50 @@ class FunctionCatalog {
+      case tf if classOf[TableFunction[T]].isAssignableFrom(tf) =>
+        Try(UserDefinedFunctionUtils.instantiate(tf.asInstanceOf[Class[TableFunction[T]]])) match {
+          case Success(tableFunction) => {
+            val clazz: Type = tableFunction.getClass.getGenericSuperclass
```
As I said earlier this and following lines are very error-prone. We should keep calls to the TypeExtractor to a very minimum. The TypeExtractor is only intended for Java not Scala.

wuchong replied:
Yes, the type extractor should be improved. `lookupTableFunction` is called only for Java users. Scala users will not go into this function.

wuchong commented on flink-libraries/flink-table/pom.xml:
```diff
@@ -154,6 +154,10 @@ under the License.
 					<execution>
+						<phase>package</phase>
+						<goals>
+							<goal>shade</goal>
```
Yes. #2673 is blocked.  Without this, my IDEA Editor couldn't resolve the shade configuration, so I introduce it in this PR. Should we remove this ? 

![image](https://cloud.githubusercontent.com/assets/5378924/20417319/757a9128-ad7f-11e6-99d7-1dc603c425a8.png)

twalthr replied:
If you have tested that a Table API word count can still be executed on the cluster, I'm fine with this change.

wuchong commented:
Thank you @fhueske @twalthr for the review, I will update the PR in this weekend.

wuchong commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/java/table/StreamTableEnvironment.scala:
```diff
@@ -164,4 +167,24 @@ class StreamTableEnvironment(
+    * @param name The name under which the function is registered.
+    * @param tf The TableFunction to register
+    */
+  def registerFunction[T](name: String, tf: TableFunction[T]): Unit ={
```
The `registerFunction[T: TypeInformation](name: String, tf: TableFunction[T])` only works in Scala. For Java, we need `registerFunction[T](name: String, tf: TableFunction[T])` and help users to extract the type. I don't expect that Java users can see `registerFunction[T: TypeInformation](name: String, tf: TableFunction[T])`.  So I separate this two function into Java/Scala TableEnvironment

wuchong commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/codegen/CodeGenerator.scala:
```diff
@@ -542,11 +563,14 @@ class CodeGenerator(
+    generateInputAccess(input._1, input._2, index, input._3)
+  }
+
+  override def visitFieldAccess(rexFieldAccess: RexFieldAccess): GeneratedExpression = {
```
sure

wuchong commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/functions/utils/TableSqlFunction.scala:
```diff
@@ -0,0 +1,99 @@
+import org.apache.flink.api.table.functions.TableFunction
+import org.apache.flink.api.table.FlinkTypeFactory
+
+import scala.collection.JavaConversions._
```
ok

wuchong commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/functions/utils/TableSqlFunction.scala:
```diff
@@ -0,0 +1,99 @@
+    for (o <- functionImpl.getParameters) {
+      val relType: RelDataType = o.getType(typeFactory)
+      argTypes.add(relType)
+      typeFamilies.add(Util.first(relType.getSqlTypeName.getFamily, SqlTypeFamily.ANY))
```
sure

wuchong commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/java/table/BatchTableEnvironment.scala:
```diff
@@ -162,4 +165,24 @@ class BatchTableEnvironment(
+  def registerFunction[T](name: String, tf: TableFunction[T]): Unit ={
+    val clazz: Type = tf.getClass.getGenericSuperclass
+    val generic = clazz match {
+      case cls: ParameterizedType => cls.getActualTypeArguments.toSeq.head
```
I will look into this.

wuchong commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/call.scala:
```diff
@@ -0,0 +1,169 @@
+    val fieldIndexes = getFieldInfo(resultType)._2
+    val function = new FlinkTableFunctionImpl(resultType, fieldIndexes, fieldNames, evalMethod)
+    val typeFactory = relBuilder.getTypeFactory.asInstanceOf[FlinkTypeFactory]
+    val sqlFunction = TableSqlFunction(tableFunction.toString, tableFunction,
```
sure

wuchong commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/table.scala:
```diff
@@ -611,6 +612,130 @@ class Table(
+        new Table(
+          tableEnv,
+          Join(this.logicalPlan, udtf.validate(tableEnv), joinType, None,
+               Some(relBuilder.getCluster.createCorrel())).validate(tableEnv))
```
Good idea!

wuchong commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/functions/utils/UserDefinedFunctionUtils.scala:
```diff
@@ -162,24 +191,107 @@ object UserDefinedFunctionUtils {
   }
 
   /**
+    * Internal method of [[ScalarFunction#getResultType()]] that does some pre-checking and uses
```
This method is never used. I carelessly introduced this method before. So I will remove this.

wuchong commented:
Regarding to the mixing parser Expression and Logical Node, how about to create a `LogicalParser` which is used to parse string to LogicalNode ? This can separate expressions and logical nodes, and keep aliases only work after table function call. 

What do you think ? @twalthr @fhueske

fhueske commented:
Hi @wuchong, I think @twalthr's main concerns are not about the parser, but rather that expressions and logical nodes are mixed, especially in the Scala Table API where the `TableFunction` is implicitly converted into a `LogicalNode` and not into an `Expression`. 
If I got it right, @twalthr would like to implicitly convert the `TableFunction` into an `Expression` at the API level and later internally into a `LogicalNode`.

Please correct me if I didn't get that right @twalthr.

wuchong commented:
Hi @fhueske @twalthr , I have addressed all the comments and made the following changes:

1. Forbid TableFunction implemented by Scala object, since the `collect` is called on a singleton. It will be error-prone in some concurrent cases.
2. Check correct errors if a SQL query refers to a function in FROM that has not been registered or which is a ScalarFunction.
3. Make `TableFunction` and `ScalarFunction` clean, **remove `UserDefinedFunction`**, and move eval relative functions and `createSqlFunction` to utils.
4. Rename `ScalarFunctions` to `SqlFunctions` because contains also TableFunction logic.
5. Restructure tests. Test Java Table API by comparing the RelNode of two tables. And check SQL API's DataSetRel and DataStreamRel via `TableTestBase` utils. The tests reduced into `stream/UserDefinedTableFunctionITCase`, `batch/UserDefinedTableFunctionITCase`, and `stream/UserDefinedTableFunctionTest` , `batch/UserDefinedTableFunctionTest`.
6. Scala Table API implicitly convert `TableFunction` into `TableFunctionCall`. `TableFunctionCall` is not an `Expression` or `LogicalNode`, but is like `GroupWindow`, can be visible to the users (contains API such `as(...)`).
7. Fix the hierarchy type extraction problem.
8. Rebase the code and fix conflicts.
9. ...

wuchong commented:
Hi @fhueske  @twalthr , could you have a look at this PR again ? I have fix the conflicts again and squashed the commits.

twalthr commented:
I have already reviewed half of the new code. I will review the rest tomorrow.

twalthr commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/expressions/ExpressionParser.scala:
```diff
@@ -451,6 +452,28 @@ object ExpressionParser extends JavaTokenParsers with PackratParsers {
 
+  lazy val tableFunctionCall: PackratParser[LogicalNode] =
+    functionIdent ~ "(" ~ repsep(expression, ",") ~ ")" ^^ {
+    case name ~ _ ~ args ~ _ => UnresolvedTableFunctionCall(name.toUpperCase, args)
```
I think it would not hurt if we support `split(c) as (f1, f2)` for all expressions. This could also help in cases where we expand the projection like `pojo.flatten() as (f1, f2)`. I don't like the idea of a logical parser. It is already confusing enough for the user that we have Java API expression strings and SQL strings, we don't need another string syntax.
@fhueske what do you think?

wuchong commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/codegen/CodeGenerator.scala:
```diff
@@ -334,18 +337,33 @@ class CodeGenerator(
+
+    val input2AccessExprs = input2 match {
+      case Some(ti) => for (i <- 0 until ti.getArity)
+        yield generateFieldAccess(ti, input2Term, i, input2PojoFieldMapping)
```
Because `generateInputAccess` will put the field access code into `reusableInputUnboxingExprs`, it is forced to be put on the top of function body. But the table function's output is an Iterable, we need to access the field in a while loop. That's why we use `generateFieldAccess` to manually manipulate the access field code.

wuchong commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/functions/TableFunction.scala:
```diff
@@ -0,0 +1,121 @@
+    *
+    * @return [[TypeInformation]] of result type or null if Flink should determine the type
+    */
+  def getResultType: TypeInformation[T] = null
```
I removed `getParameterTypes()` from `TableFunction` intentionally. Because it is never used in current implementation, and it is a little complex and needs sufficient tests to support it. I would like to support it in next JIRA, and I created [FLINK-5225](https://issues.apache.org/jira/browse/FLINK-5225) just now.

This needs to implement a custom Calcite's TableFunction and override `getParameters()`. But currently, the `FlinkTableFunctionImpl` extends `ReflectiveFunctionBase` which determines the parameter types of the eval method.

wuchong commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/TableFunctionCall.scala:
```diff
@@ -0,0 +1,110 @@
+}
+
+
+case class TableFunctionCallBuilder[T: TypeInformation](udtf: TableFunction[T]) {
```
I would move this into `ExpressionUtils`, learned from your new [PR](https://github.com/apache/flink/pull/2919/files#diff-04d1bca648d7ee47ab9ce787c8d944a6R567)  😄

wuchong commented:
Hi @fhueske @twalthr ,  I have updated the PR, please review it again when you are available.

twalthr commented:
@wuchong thanks for updating the PR! It looks good now. I will rebase it, quickly scan over the code again and merge it tomorrow.

However, we should fix FLINK-5223 as soon as possible. It would be great if someone could do that before the 1.2 release.

wuchong commented:
Thanks @twalthr , I will add the documentation ASAP.

twalthr commented:
Thanks. Merging...
